src/dynamic_bridge.cpp

```diff
@@ -500,319 +509,592 @@ int main(int argc, char * argv[])
+          std::lock_guard<std::mutex> lock(g_bridge_mutex);
+          ros1_services = active_ros1_services;
         }
 
-        auto publisher_count = ros2_node->count_publishers(topic_name);
-        auto subscriber_count = ros2_node->count_subscribers(topic_name);
+        // get message types for all topics
+        ros::master::V_TopicInfo topics;
+        bool success = ros::master::getTopics(topics);
+        if (!success) {
+          fprintf(stderr, "failed to poll ROS 1 master\n");
+          return;
+        }
 
-        // ignore publishers from the bridge itself
-        if (bridges_1to2.find(topic_name) != bridges_1to2.end()) {
-          if (publisher_count > 0) {
-            --publisher_count;
+        std::map<std::string, std::string> current_ros1_publishers;
+        std::map<std::string, std::string> current_ros1_subscribers;
+        for (auto topic : topics) {
+          auto topic_name = topic.name;
+          bool has_publisher = active_publishers.find(topic_name) != active_publishers.end();
+          bool has_subscriber = active_subscribers.find(topic_name) != active_subscribers.end();
+          if (!has_publisher && !has_subscriber) {
+            // skip inactive topics
+            continue;
           }
-        }
-        // ignore subscribers from the bridge itself
-        if (bridges_2to1.find(topic_name) != bridges_2to1.end()) {
-          if (subscriber_count > 0) {
-            --subscriber_count;
+          if (has_publisher) {
+            current_ros1_publishers[topic_name] = topic.datatype;
+          }
+          if (has_subscriber) {
+            current_ros1_subscribers[topic_name] = topic.datatype;
+          }
+          if (output_topic_introspection) {
+            printf(
+              "  ROS 1: %s (%s) [%s pubs, %s subs]\n",
+              topic_name.c_str(), topic.datatype.c_str(),
+              has_publisher ? ">0" : "0", has_subscriber ? ">0" : "0");
           }
         }
 
-        if (publisher_count) {
-          current_ros2_publishers[topic_name] = topic_type;
+        // since ROS 1 subscribers don't report their type they must be added anyway
+        for (auto active_subscriber : active_subscribers) {
+          if (current_ros1_subscribers.find(active_subscriber) == current_ros1_subscribers.end()) {
+            current_ros1_subscribers[active_subscriber] = "";
+            if (output_topic_introspection) {
+              printf("  ROS 1: %s (<unknown>) sub++\n", active_subscriber.c_str());
+            }
+          }
         }
 
-        if (subscriber_count) {
-          current_ros2_subscribers[topic_name] = topic_type;
+        if (output_topic_introspection) {
+          printf("\n");
         }
 
-        if (output_topic_introspection) {
-          printf(
-            "  ROS 2: %s (%s) [%zu pubs, %zu subs]\n",
-            topic_name.c_str(), topic_type.c_str(), publisher_count, subscriber_count);
+        {
+          std::lock_guard<std::mutex> lock(g_bridge_mutex);
```

Comment:
You may have already found this in testing, but should this block be part of the previous if-statement?